### DISCUSSION TITLE
13.0.10: DatePicker not updating on select; viewChange event not updating model; selectOtherMonths not working

### DISCUSSION THREAD

**Discussion by jncolin** [primefaces/community#2137] [PrimeFaces] [answered]
Hi, 

I'm using PrimeFaces 13.0.10 with MyFaces 4.0.2. I have a page with a DatePicker component, and I'm facing strange problems, that did not happen in preceding versions. The DatePicker must allow multiple selection.

```
        <h:form>
            <p:datePicker id="dateField" inline="true" widgetVar="wgtDatePicker"
                          selectionMode="multiple" readonlyInput="true" selectOtherMonths="false" 
                          value="#{playGround.selectedDates}" showButtonBar="true">
                <p:ajax event="dateSelect" 
                        listener="#{playGround.dateSelectListener}"
                        update="message"/>
                <p:ajax event="viewChange" 
                        listener="#{playGround.dateViewChangeListener}"
                        update="message"/>
            </p:datePicker>
            <br />
            <h:outputText id="message">#{playGround.message}</h:outputText>
        </h:form>
```

The backing bean looks like this:
```
@Named
@ViewScoped
public class PlayGround implements Serializable {

    private String message;
    private List<LocalDate> selectedDates;
    private final Logger logger = Logger.getLogger(getClass().getName());

    public PlayGround() {
        message = "Hello!";
    }

    public String getMessage() {
        logger.log(Level.FINE, "getMessage: {0}", message);
        return message;
    }

    public List<LocalDate> getSelectedDates() {
        logger.log(Level.FINE, "getSelectedDates: {0}", selectedDates);
        return selectedDates;
    }

    public void setSelectedDates(List<LocalDate> selectedDates) {
        logger.log(Level.FINE, "setSelectedDates: {0}", selectedDates);
        this.selectedDates = selectedDates;
    }

    public void dateSelectListener(SelectEvent event) {
        logger.log(Level.FINE, "dateSelectListener: {0}", selectedDates);
        message = "dateSelectListener: " + event.getObject();
    }

    public void dateViewChangeListener(DateViewChangeEvent event) {
        logger.log(Level.FINE, "dateViewChangeListener: {0}", selectedDates);
        message = String.format("dateViewChangeListener: %d/%d", event.getMonth(), event.getYear());
    }
}
```

### First problem: date select/unselect
I'm using a Ajax DateSelect Listener that fires when I select a Day in the calendar. It runs properly and updates the underlying array in the model when I select a Date. But if I un-select a day that was previously selected, it does not fire, hence the array is not updated. If I select a new day, then the un-selected day disappears from the array.

### Second problem: selectOtherMonths
Eventhough I specified selectOtherMonths="false", it is still possible to select dates in different months.

### Third problem: viewChange
I have created a listener for the viewChange event; If the event handler is defined like this:
```
                <p:ajax event="viewChange" 
                        listener="#{playGround.dateViewChangeListener}"
                        update="message"/>
```
it works fine. But if I change it to
```
                <p:ajax event="viewChange" 
                        listener="#{playGround.dateViewChangeListener}"
                        update="@form"/>
```
Then it becomes impossible to change the view; it always comes back to the current month, as if the underlying date was not updated

### Fourth and last problem: message is concatenated instead of being updated
Whenever a listener is run that updates the message field of the bean, the new value of the field appears in the webpage as concatenated to the previous value. I can't get why it is so.

I'll be really grateful if anyone could help me address these problems

Thanks in advance

Jean

**Comment by melloware** [comment]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

That would help me see your issues.  Also have you tried 14.0.1 to see if these issues are already resolved?

**Reply by jncolin** [reply]
Here's a tar file with the project including my test file: head to http://localhost:9080/playground.xhtml
[primefaces-test-jakarta.zip](https://github.com/user-attachments/files/15838778/primefaces-test-jakarta.zip)

**Reply by melloware** [reply]
can you tell me the last version of PrimeFaces where this code worked?

**Reply by melloware** [reply]
@jncolin try this its all working for me after some tweaks to your code.

[discussion-2137.zip](https://github.com/user-attachments/files/15839339/discussion-2137.zip)

I can't reproduce ANY of your issues above. run `mvn clean jetty:run` and off you go...

**Comment by jncolin** [comment]
@melloware I tried on my side, simply unzipping and running mvn clean jetty:run

1. Regarding, select/unselect, it indeed works, I don't understand why it doesn't in my case
2. Regarding viewChange, I confirm the issue; if the update attribute in the viewChange ajax tag is 'message', I can change month; if it is set to @form, I can't, it always jumps back to the current month
3. Regarding selecting dates in the current month only, if I set the 'update' attribute to 'message', I can change month in the view; but I can select dates from any months, and these will simply add to the list displayed
4. Regarding message concatenation, you are right, it works, I still have to find out why it doesn't on my side.

I have simple use cases:
- restrict date selection to the currently displayed month only, even if dates have been previously selected in other months before the view changed
- when the view is changed, I need to invoke a bean method to recalculate some other elements in the form
- when a method is completed in the bean after clicking submit, I want the displayed month to remain the same and not reset to today's month. 
I understand there are two variables: the selectedDates (Array in case of multi selection) and the ViewDate (that controls the currently viewed month).
I really appreciate your help
Best regards

**Comment by jncolin** [comment]
To help you understand, here are two screen recording I made. The first one is with the update attribute set to @form, the second one is with "message"

As you can see, in the first one, I can't change month, and the message gets longer and longer; in the second one, I can change month and select dates from any months; again, the message gets longer and longer

Thanks a lot


https://github.com/primefaces/community/assets/3629802/7a00c337-ada9-4b74-8428-ace62641defe


https://github.com/primefaces/community/assets/3629802/dd930069-0c4d-4077-93a9-8409222a1b3c

**Reply by melloware** [reply]
OK screen recordings helped.  Let me look again.

**Reply by melloware** [reply]
@jncolin can you answer my question you said this code used to work what version of PrimeFaces can i test where this is working?  because the `selectOtherMonths` code has not changed in 5 years according to Git.

**Reply by melloware** [reply]
OK got to the bottom of another mystery. You are misunderstanding what `selectOtherMonths="false"` means.  Let me explain with screenshots.

### selectOtherMonths=false and showOtherMonths=false

![image](https://github.com/primefaces/community/assets/4399574/5501074f-aae0-4754-b898-d96bb3aaec79)

### selectOtherMonths=false and showOtherMonths=true

![image](https://github.com/primefaces/community/assets/4399574/9413c1a1-def3-4d71-82b2-24222aa818bf)

### selectOtherMonths=true and showOtherMonths=true

![image](https://github.com/primefaces/community/assets/4399574/a1cc6d9b-e23e-48f8-9d30-4b70c337e775)

**Reply by melloware** [reply, marked as answer]
Solved more of your issues. try this version:
[discussion3-2137.zip](https://github.com/user-attachments/files/15842366/discussion3-2137.zip)

- [x] Dates in other months are now not selectable
- [x] Update works with `@obs` instead of `@form`

Observer pattern in action: https://www.primefaces.org/showcase/ui/ajax/observer.xhtml?

**Reply by jncolin** [reply]
Thanks for this; actually, it does not really solve my issue, because it limits the current input dates to the current month. Maybe I need to clarify my scenario: We are in June, so the displayed month is june; let's say I select 3rd of June, but then, I change view because I realize that the correct dates to select is not June 3rd but May 10 and 13; so I click the 'previous' button, get on May, and there I select 10 and 13. But I want to remove June 3rd from the list, because of the month switch. One way I found to do that is to call PF('wgtDatePicker1').setDate(null); in the oncomplete event of the viewChange ajax event. Is this the right way? 
And then, when I submit the form, how can I return to May, but with no date selected? 
Thanks for your support

**Comment by jncolin** [comment]
I think the issue can be closed. Basically, beyond the other month select I misunderstood, my main issue is that I can’t find a way to set a date picker view AND selected dates at the same time from the bean, if no date is selected. If a date is selected, value in date picker contains the list of selected dates and the month is displayed accordingly. But if there is no date selected, then I cannot set the month view programmatically. Anyway thanks for your help @melloware

**Reply by melloware** [reply]
@jncolin someone else just requested your `initialDate` feature: https://github.com/primefaces/primefaces/issues/12157
